### DISCUSSION TITLE
Octane - computed properties don't get reset between requests

### DISCUSSION THREAD

**Discussion by lalov** [livewire/livewire#3043] [7. Bugs (archived)]
### Description
When using computed properties on a component, the method that computes the property only seems to execute once per swoole instance. 

### A fix?
This seems to fix it, but I don't know where to put it in a pull request: 

<pre>
        Livewire::listen('component.hydrate', function ($component, $request) {
            $component->forgetComputed();
        });
</pre>

### Exact steps to reproduce
Create a component with a computed property, and load the component from a route. 
Open the page so that the computed property gets different values, and reload the page multiple times. 
The page will show "old" values in random order.

**Context**
- Livewire version: 2.4.2
- Laravel version: 8.37.0

**Comment by calebporzio** [comment]
Perfect fix, thanks! 

If you could PR this that'd be great.

I'd put it in a dedicated hydration middleware and reference in the service provider.

Hope that helps, thanks again!

**Comment by joshhanley** [comment]
@calebporzio interesting that the computed properties aren't cleared.

I haven't played with it yet, but does that mean the components don't get destroyed at the end of a lifecycle? 

I wonder what other side effects there are if thats the case? Should the new middleware actually deconstruct a component when the lifecycle is finished, or is there a benefit that it doesn't get destroyed?

**Comment by lalov** [comment]
@joshhanley I checked, and the constructor for the component only runs for the first requests, so it doesn't get destroyed. 
I think it matters that the component is run from as a full-page-component.

**Comment by lalov** [comment]
Another way to fix it is to reset computed properties in the __invoke() method in Livewire/Component

<pre>

public function __invoke(Container $container, Route $route)
{
    $this->computedPropertyCache = [];

    $componentParams = (new ImplicitRouteBinding($container))
    ->resolveAllParameters($route, $this);

</pre>

**Comment by joshhanley** [comment]
@lalov ah right, thanks for the info 🙂 

Hmm so yeah I do wonder then whether it would be better to add a middleware that actually deconstructs the components rather than leave them running. I could imagine issues if data is assigned to any protected/ private properties on a component.

That would then also deal with the computed property issue.

**Comment by calebporzio** [comment]
I agree Josh. Components should be destroyed after we're done with them.

**Comment by lalov** [comment]
I agree, it would be good to deconstruct the components. I dont think I'll be able to make the PR...

**Comment by ju5t** [comment]
I can help with the PR if someone could give me some pointers where to look for :)

This might solve an issue I'm having with query strings leaking to another component.

**Comment by joshhanley** [comment]
@ju5t have a look in the `LivewireServiceProvider`, you will see a section on hydration middlewares. 

There is a stack for subsequent requests and also a stack for initial dehydration.

A middleware here would probably be the best, that runs as the last step of dehydration, in both of those stacks, that deconstructs the instance.

If that doesn't work, it will probably need to be done inside the `LifecycleManager` at the end of the `initialDehydrate` and `dehydrate` methods.

I'm not running Octane at this stage, so I personally have no way of testing if it would work or not, but happy to support you with getting it going.

Hope this helps!

**Comment by ju5t** [comment]
@joshhanley the source of Livewire is still a bit overwhelming for me personally :)

I made this `DeconstructComponent` middleware:

```php
<?php

namespace Livewire\HydrationMiddleware;

class DeconstructComponent implements HydrationMiddleware
{
    public static function hydrate($instance, $request)
    {
        $instance->forgetComputed();
    }

    public static function dehydrate($instance, $response)
    {
        $instance->forgetComputed();
    }
}
```

I have added it above `HydratePublicProperties::class` in

- `registerHydrationMiddleware` before `HydratePublicProperties` (as it runs in reverse order when dehydrating, right?) I _think_ this is wrong place somehow.
- `registerInitialDehydrationMiddleware` below `NormalizeComponentPropertiesForJavaScript`

This doesn't solve things for me but just checking if I'm on the right path here. I'm not even sure what to do to properly deconstruct the component. I clearly lack some knowledge on the internals of Livewire.

**Comment by 3adeling** [comment]
I don't see any open issues related to this problem. is it fixed or not yet?

**Reply by ju5t** [reply]
This used to be the issue but it got moved to discussions. It's not fixed. I did some work on a PR but got stuck. I will either need some help, or someone else should make the PR.

**Comment by joshhanley** [comment]
This has been fixed in https://github.com/livewire/livewire/pull/3947 and should be in the next release of Livewire.

Hope this helps!
